### DISCUSSION TITLE
GCNConv Dimension out of range (expected to be in range of [-1, 0], but got -2)

### DISCUSSION THREAD

**Discussion by JoseDiogoMarquesdosSantos** [pyg-team/pytorch_geometric#9252] [Q&A] [answered]
Hello.
I'm trying to create a model for a graph classification task. I have 400 graphs for training and another 200 for testing, every graph has the same adjacency matrix (99x99 with binary values) and every node has 1 attribute that is a real number. The graphs are created according to
```
for i in range(len(train_in)):
    x = torch.from_numpy(train_in[i])
    edge_index = torch.nonzero(torch.from_numpy(adj_matrix), as_tuple=False).t().contiguous()
    y =  torch.from_numpy(train_out[i])
    graph = Data(x=x, edge_index=edge_index, y=y)
    train_graphs.append(graph)
```
And similarly for the test group, I then save each group in a .pt file and import them according to
```
train_data = torch.load("train_set.pt")
test_data = torch.load("test_set.pt")
train_loader = DataLoader(train_data, shuffle=False)
test_loader = DataLoader(test_data, shuffle=False)
```
I then define a custom activation function and the model itself
```
class tansig(torch.autograd.Function):
    @staticmethod
    def forward(ctx, input, a, b):
        # Calculate the forward pass
        result = a * torch.tanh(input * b)
        ctx.save_for_backward(input, a, b)
        return result
    @staticmethod
    def backward(ctx, grad_output):
        # Calculate the backward pass
        input, a, b = ctx.saved_tensors
        grad_input = grad_output * a * b * (1 - torch.tanh(input * b)**2)
        grad_a = torch.sum(grad_output * torch.tanh(input * b))
        grad_b = torch.sum(grad_output * a * input * (1 - torch.tanh(input * b)**2))
        return grad_input, grad_a, grad_b
class Tansig(nn.Module):
    def __init__(self, a, b):
        super(Tansig, self).__init__()
        self.a = nn.Parameter(torch.tensor(a, dtype=torch.float32))
        self.b = nn.Parameter(torch.tensor(b, dtype=torch.float32))
    def forward(self, input):
        return tansig.apply(input, self.a, self.b)
class GNNModel(nn.Module):
    def __init__(self, n_inputs):
        super(GNNModel, self).__init__()
        self.conv1 = GCNConv(-1, n_inputs)
        self.layer1 = nn.Linear(n_inputs, 10)
        self.output = nn.Linear(10, 5)
        self.tansigAct = Tansig(a = 1/np.tanh(2/3), b = 2/3)
        self.sigmoid = nn.Sigmoid()
    def forward(self, x, edge_index):
        # Apply first convolutional layer
        x = self.conv1(x, edge_index)
        x = F.relu(x)
        # Apply global mean pooling
        x = global_mean_pool(x)
        # Apply final layers
        x = self.layer1(x)
        x = self.tansigAct(x)
        x = self.output(x)
        x = self.sigmoid(x)
        return x
model = GNNModel(n_inputs=99)
criterion = nn.CrossEntropyLoss()
optimizer = torch.optim.Adam(model.parameters(), lr=0.001)
```
And I attempt to train it according to
```
num_epochs = 10
for epoch in range(num_epochs):
    model.train()
    for data in train_loader:
        optimizer.zero_grad()
        out = model(data.x, data.edge_index)
        loss = criterion(out, data.y)
        loss.backward()
        optimizer.step()
```
However I get a "Dimension out of range (expected to be in range of [-1, 0], but got -2)" error related to the line "out = model(data.x, data.edge_index)", in particular to the edge_index part as indicated by "self._cached_edge_index = (edge_index, edge_weight)" in the error message. The full error message is:
"Cell In[165], line 7
      5 for data in train_loader:
      6     optimizer.zero_grad()
----> 7     out = model(data.x, data.edge_index)
      8     loss = criterion(out, data.y)
      9     loss.backward()

File c:\Users\***\AppData\Local\Programs\Python\Python310\lib\site-packages\torch\nn\modules\module.py:1518, in Module._wrapped_call_impl(self, *args, **kwargs)
   1516     return self._compiled_call_impl(*args, **kwargs)  # type: ignore[misc]
   1517 else:
-> 1518     return self._call_impl(*args, **kwargs)

File c:\Users\***\AppData\Local\Programs\Python\Python310\lib\site-packages\torch\nn\modules\module.py:1527, in Module._call_impl(self, *args, **kwargs)
   1522 # If we don't have any hooks, we want to skip the rest of the logic in
   1523 # this function, and just call forward.
   1524 if not (self._backward_hooks or self._backward_pre_hooks or self._forward_hooks or self._forward_pre_hooks
   1525         or _global_backward_pre_hooks or _global_backward_hooks
   1526         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1527     return forward_call(*args, **kwargs)
   1529 try:
   1530     result = None

Cell In[163], line 12
...
    224         self.improved, self.add_self_loops, self.flow, x.dtype)
    225     if self.cached:
    226         self._cached_edge_index = (edge_index, edge_weight)

IndexError: Dimension out of range (expected to be in range of [-1, 0], but got -2)"
I have checked and the dimension of both data.x and data.edge_index seems to be correct a 1x99 vector for x and a 2x733 matrix for edge_index (there's only 733 connections in the graph). The edges are unweighted (hence the adjacency matrix being binary values). The output is a 1x5 1-hot vector. I have tried every troubleshooting method I have thought of and seen online but nothing as worked and I don't even fully understand what the error means. Could this error be caused by some of the nodes being unconnected to the rest of the graph? If not, what could be causing it?
Thanks

**Comment by rusty1s** [comment, marked as answer]
What is the shape of `data.x`? Note that even for one-dimensional features, it needs to be encoded as `[num_nodes, 1]`.

**Reply by JoseDiogoMarquesdosSantos** [reply]
That was it, running data.x.size() returned torch.Size([99]). By adding data.x.unsqueeze(1) it no longer throws that error. Thanks
